packages/rule-no-broken-links/src/rule.ts

```diff
@@ -0,0 +1,135 @@
+            const statusIndex = brokenStatusCodes.indexOf(
+                networkData.response.statusCode
+            );
+
+            if (statusIndex > -1) {
+                await context.report(
+                    absoluteUrl,
+                    linkElement,
+                    `Broken link found (${
+                        brokenStatusCodes[statusIndex]
+                    } response)`
+                );
+            }
+        };
+
+        // This method checks whether a url is broken and report if it is
+        const reportBrokenLink = async (hrefUrl: string, resource: any, linkElement: IAsyncHTMLElement) => {
+            const absoluteUrl = new URL(hrefUrl, resource).href;
+
+            await requester
+                .get(absoluteUrl)
+                .then((value: NetworkData) => {
+                    handleSuccess(value, absoluteUrl, linkElement);
+                })
+                .catch((error: any) => {
+                    handleRejection(error, absoluteUrl, linkElement);
+                });
+        };
+
+        const checkImageElement = async (linkElement: IAsyncHTMLElement, resource: any) => {
+            const imageSrc = linkElement.getAttribute('src');
+
+            await reportBrokenLink(imageSrc, resource, linkElement);
+
+            // Check for srcset attribute value if that exist
+            const srcSet = linkElement.getAttribute('srcset');
+
+            if (srcSet !== null) {
+                const sources = srcSet.split(',').filter((x: string) => {
+                    return x !== '';
+                });
+
+                for (const source of sources) {
+                    const url = source.split(' ').filter((x: string) => {
+                        return x !== '';
+                    })[0];
+
+                    await reportBrokenLink(url, resource, linkElement);
+                }
+            }
+        };
+
+        const checkAnchorTag = async (linkElement: IAsyncHTMLElement, resource: any) => {
+            const hrefUrl = linkElement.getAttribute('href');
+
+            if (hrefUrl !== '/') {
+                await reportBrokenLink(hrefUrl, resource, linkElement);
+            }
+        };
+
+        const validateElement = async (elementFound: ElementFound) => {
+            const { resource } = elementFound;
+
+            debug(`Validating rule no-broken-links`);
+
+            const linkElements = await elementFound.element.ownerDocument.querySelectorAll(
+                'a,img'
+            );
+
+            for (let i = 0; i < linkElements.length; i++) {
+                const linkElement = linkElements[i];
+
+                if (linkElement.nodeName === 'IMG') {
+                    await checkImageElement(linkElement, resource);
+                } else {
+                    await checkAnchorTag(linkElement, resource);
+                }
+            }
+        };
+
+        context.on('element::body', validateElement);
```

Comment:
The way we are doing it with other rules is to subscribe to `element::a`, `element::img`, instead of using `querySelectorAll`.
There are a few other things we should be checking like scripts, styles, video elements, posters, etc. If they are in the critical path they will slow even more the loading (especially if they timeout).
The rule [https-only](https://github.com/sonarwhal/sonarwhal/blob/4771017d6029828265d96377cdc433a1229add3b/packages/rule-https-only/src/rule.ts#L183) should have all the elements that need to be analyzed (except anchor tags) and how to get the `src`/`srcset`/`href` in a generic way. I have a concern with the videos if they are using adaptive streaming, drm or something like that. I'm not sure what the answer from the server will be. We should check it.

Also the description of the rule (in comments, etc.) should be updated. Maybe use something generic like:

> Verify that all links and resources the page uses are available online.